server/WAPI.rb

```diff
@@ -90,56 +83,89 @@ def self.base64_key_secret(key, secret)
     Base64.strict_encode64(ks)
   end
 
-  # use instance information to generate the proper url
+  ######### utilities for URL formatting
+  # use instance specific configuration information to generate the full url.
   def format_url(request)
     "#{@api_prefix}#{request}"
   end
 
-  def do_request(request)
+  # Responses may contain partial results.  In that case information about how to get the remaining data is returned
+  # in the 'Link' header. Link headers come back with explicit URLs pointing to Canvas servers.
+  # Remove that server added by the external service since we need to send queries back through the ESB rather than
+  # straight to the external service.
+  # This code assumes that the main portion of the next link (more_url) looks like the original request except that
+  # the beginning may be different (e.g. direct to canvas host vs through the ESB proxy) and the query parameters
+  # may be different (e.g. maybe add info about where to restart retrieval).
+
+  # A trivial more_url should become an empty string.
+  # Query parameters on the more_url should be passed through.
+  # Query parameters on the request_string should be ignored.
+  # The more_url is a complete URL that was returned from the external service.  The
+  # request_string it the partial url generated internally that doesn't have any explicit server / api
+  # version information.
+
+  def reduce_url(more_url, request_string)
+    logger.debug "#{self.class.to_s}:#{__method__}:#{__LINE__}: more_url: #{more_url} request_string: #{request_string}"
+
+    return "" if more_url.nil? || more_url.length == 0
+
+    # Get the main part of the original request without any query parameters.
+    main_request = request_string.index('?') ? request_string[/(^.+)\?/] : request_string
+
+    # Pull out the part of more_url that matches the original query and also pull along query parameters that were
+    # supplied in the more_url.
+    more_url = more_url[/#{main_request}.*/]
+
+    logger.debug "#{self.class.to_s}:#{__method__}:#{__LINE__}: reduced more_url: #{more_url}"
+    more_url
+  end
+
+  # Make a request to an external service and handle error conditions and headers.
+  def do_request(request_string)
 
     RestClient.log = logger if (logger.debug? and TRACE != FalseClass)
 
-    url=format_url(request)
+    # make the request specific to the separately configured API.
+    url=format_url(request_string)
```

Comment:
It doesn't appear to be.  I would expect `request` alone to be a request object.  So, `request_string` could be the string representation of that object, including headers, cookies, etc., like:

```
GET /tl-its-umich-edu/StudentDashboard/pull/303/files HTTP/1.1\nHost: github.com\nConnection: keep-alive\nPragma: no-cache\nCache-Control: no-cache\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\nUpgrade-Insecure-Requests: 1\nUser-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36\nAccept-Encoding: gzip, deflate, sdch\nAccept-Language: en-US,en;q=0.8\nCookie: _octo=GH1.1.1815.140534; logged_in=yes; dotcom_user=lsloan; _gat=1; _ga=GA1.2.35724.13527; tz=America%2FDetroit; _gh_sess=2e08320383c09f596f11e99170; user_session=6qb7MhuPCL\n
```